### PR TITLE
Track platform in MigratePlaformConfig and enforce metadata file being provided

### PR DESCRIPTION
**Pull request by GaryQian**
Tracks the platform in MigratePlaformConfig so that similar instances for different platforms don't equal each other.

Enforce FlutterProjectMetdata has a file associated with it.

### CODE REVIEW

**Review by christopherfujino — COMMENTED** [review]
Does this factory need to live in this file, versus being a standalone function that returns a `FlutterProjectMetadata.explicit()`? It seems to me like this logic is specific to Flutter migrate, and if you needed to tweak it, there would need to make two PRs rather than 1 (the first PR updating the factory logic in this file, the second to roll that change to flutter migrate).

File: packages/flutter_tools/lib/src/flutter_project_metadata.dart [hunk 1]
```diff
@@ -141,16 +141,27 @@ class FlutterProjectMetadata {
   final Logger _logger;
 
   final File _metadataFile;
+  File get file => _metadataFile;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/flutter_project_metadata.dart:144]: should `_metadataFile` just be final public then?

Comment (GaryQian) [reply packages/flutter_tools/lib/src/flutter_project_metadata.dart:144]: Yeah, I guess so!

**Review by christopherfujino — APPROVED** [review]
LGTM!

**Pull request merged**